pkg/maintainer/bitcoin_difficulty.go

```diff
@@ -0,0 +1,347 @@
+package maintainer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-log"
+	"github.com/keep-network/keep-core/pkg/bitcoin"
+)
+
+const (
+	// Default value for back-off time which should be applied when the Bitcoin
+	// difficulty maintainer is restarted. It helps to avoid being flooded with
+	// error logs in case of a permanent error in the Bitcoin difficulty
+	// maintainer.
+	defaultRestartBackoffTime = 120 * time.Second
+
+	// Default value for back-off time which should be applied when there are no
+	// more Bitcoin epochs to be proven because the difficulty maintainer is
+	// up-to-date with the Bitcoin blockchain or there are not enough blocks yet
+	// to prove the new epoch.
+	defaultIdleBackOffTime = 60 * time.Second
+
+	// The number of blocks in a Bitcoin difficulty epoch.
+	bitcoinDifficultyEpochLength = 2016
+)
+
+var logger = log.Logger("keep-maintainer-bitcoin-difficulty")
+
+var (
+	errNotAuthorized = fmt.Errorf(
+		"bitcoin difficulty maintainer has not been authorized to submit " +
+			"block headers",
+	)
+	errNoGenesis = fmt.Errorf(
+		"genesis has not been performed in the Bitcoin difficulty chain",
+	)
+)
+
+func initializeBitcoinDifficultyMaintainer(
+	ctx context.Context,
+	btcChain bitcoin.Chain,
+	chain BitcoinDifficultyChain,
+	idleBackOffTime time.Duration,
+	restartBackOffTime time.Duration,
+) {
+	bitcoinDifficultyMaintainer := &BitcoinDifficultyMaintainer{
+		btcChain:           btcChain,
+		chain:              chain,
+		idleBackOffTime:    idleBackOffTime,
+		restartBackOffTime: restartBackOffTime,
+	}
+
+	go bitcoinDifficultyMaintainer.startControlLoop(ctx)
+}
+
+// BitcoinDifficultyMaintainer is the part of maintainer responsible for
+// maintaining the state of the Bitcoin difficulty on-chain contract.
+type BitcoinDifficultyMaintainer struct {
+	btcChain bitcoin.Chain
+	chain    BitcoinDifficultyChain
+
+	idleBackOffTime    time.Duration
+	restartBackOffTime time.Duration
+}
+
+// startControlLoop starts the loop responsible for controlling the Bitcoin
+// difficulty maintainer.
+func (bdm *BitcoinDifficultyMaintainer) startControlLoop(ctx context.Context) {
+	logger.Info("starting Bitcoin difficulty maintainer")
+
+	defer func() {
+		logger.Info("stopping Bitcoin difficulty maintainer")
+	}()
+
+	for {
+		err := bdm.proveEpochs(ctx)
+		if err != nil {
+			logger.Errorf(
+				"restarting Bitcoin difficulty maintainer due to error while "+
+					"proving Bitcoin blockchain epochs: [%v]",
+				err,
+			)
```

Comment:
Let's make it more concise: `error while proving Bitcoin epochs: [%v]; restarting maintainer`

Reply:
Done in 1019eaa16ab83309fbd0f34124f250d91106dd7d.